File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/IDCardToolkitObjCApp/IDCardToolkitObjCApp/ViewController/NFCPublicDataViewController.m
//
//  NFCPublicDataViewController.m
//  IDCardToolkitObjCApp
//
//  Created by Federal Authority For Identity and Citizenship on 14/10/19.
//  Copyright © 2019 Federal Authority For Identity and Citizenship. All rights reserved.
//

#import "NFCPublicDataViewController.h"

@interface NFCPublicDataViewController ()
@property(nonatomic,strong)Utils * utils;
@end

@implementation NFCPublicDataViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // init objects
       
       getpModDataDictionary=[[NSMutableDictionary alloc]init];
       getpNonModDataDictionary=[[NSMutableDictionary alloc]init];
       getpHomeAddressDictionary=[[NSMutableDictionary alloc]init];
       getpWorkAddressDictionary=[[NSMutableDictionary alloc]init];
       
       pIDNumberKeyarr=[[NSMutableArray alloc]init];
       pIDNumberValuearr=[[NSMutableArray alloc]init];
       pModDatakeyarr=[[NSMutableArray alloc]init];
       pModDataValuearr=[[NSMutableArray alloc]init];
       pNonModDatakeyarr=[[NSMutableArray alloc]init];
       pNonModDataValuearr=[[NSMutableArray alloc]init];
       pHomeAddresskeyarr=[[NSMutableArray alloc]init];
       pHomeAddressValuearr=[[NSMutableArray alloc]init];
       pWorkAddresskeyarr=[[NSMutableArray alloc]init];
       pWorkAddressValuearr=[[NSMutableArray alloc]init];
       
       self.photoImage.layer.cornerRadius=8.0f;
       self.photoImage.layer.borderWidth=1.0f;
       self.photoImage.layer.masksToBounds = YES;
       self.photoImage.layer.borderColor=[UIColor clearColor].CGColor;
       self.photoImage.layer.shadowOffset = CGSizeMake(-15, 20);
       self.photoImage.layer.shadowRadius = 8;
       self.photoImage.layer.shadowOpacity = 0.1;
       
       self.signImage.layer.cornerRadius=8.0f;
       self.signImage.layer.borderWidth=1.0f;
       self.signImage.layer.masksToBounds = YES;
       self.signImage.layer.borderColor=[UIColor clearColor].CGColor;
       self.signImage.layer.shadowOffset = CGSizeMake(-15, 20);
       self.signImage.layer.shadowRadius = 8;
       self.signImage.layer.shadowOpacity = 0.1;
       
       self.title=@"NFC Public data";
       self.navigationItem.hidesBackButton = YES;
       self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[[UIImage imageNamed:@"backbutton"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal]
                                                                                style:UIBarButtonItemStylePlain
                                                                               target:self
                                                                               action:@selector(goBack)];
       
        AppDelegate *aDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
       self.model = aDelegate.model;
       self.utils=aDelegate.utils;

    // Do any additional setup after loading the view.
}
-(void)getPublicaDataClass:(CardPublicData *)cardPublicData {
    NSLog(@"NFC Details View @s",cardPublicData);
            dispatch_async(dispatch_get_main_queue(), ^{
                
                [self->pIDNumberKeyarr addObject:@"IDNumber"];
                [self->pIDNumberKeyarr addObject:@"CardNumber"];
                [self->pIDNumberValuearr addObject:[cardPublicData getIdNumber]];
                [self->pIDNumberValuearr addObject:[cardPublicData getCardNumber]];
                        
                    PublicDataParse *parseData=[[PublicDataParse alloc]init];
                self->getpModDataDictionary=[parseData getModifiablePublicDataDetails:[cardPublicData getModifiablePublicData]];
                self->getpNonModDataDictionary=[parseData getNonModifiablePublicDataDetails:[cardPublicData getNonModifiablePublicData]];
                self->getpHomeAddressDictionary=[parseData getHomeAddressDetails:[cardPublicData getHomeAddress]];
                self->getpWorkAddressDictionary=[parseData getWorkAddressDetails:[cardPublicData getWorkAddress]];
                    
                BOOL isEmpty = ([self->getpModDataDictionary count] == 0);
                    if (!isEmpty) {
                        NSArray * pModDataKeys;
                        pModDataKeys = [self->getpModDataDictionary allKeys];
                        for (NSString *key in pModDataKeys) {
                            id value = self->getpModDataDictionary[key];
                            [self->pModDatakeyarr addObject:key];
                            [self->pModDataValuearr addObject:value];
                        }// forloop
                        NSArray * pNonModDataKeys;
                        pNonModDataKeys = [self->getpNonModDataDictionary allKeys];
                        for (NSString *key in pNonModDataKeys) {
                            id value = self->getpNonModDataDictionary[key];
                            [self->pNonModDatakeyarr addObject:key];
                            [self->pNonModDataValuearr addObject:value];
                        }// forloop
                        NSArray * pHomeAddressKeys;
                        pHomeAddressKeys = [self->getpHomeAddressDictionary allKeys];
                        for (NSString *key in pHomeAddressKeys) {
                            id value = self->getpHomeAddressDictionary[key];
                            [self->pHomeAddresskeyarr addObject:key];
                            [self->pHomeAddressValuearr addObject:value];
                        }// forloop
                        NSArray * pWorkAddressKeys;
                        pWorkAddressKeys = [self->getpWorkAddressDictionary allKeys];
                        for (NSString *key in pWorkAddressKeys) {
                            id value = self->getpWorkAddressDictionary[key];
                            [self->pWorkAddresskeyarr addObject:key];
                            [self->pWorkAddressValuearr addObject:value];
                        }// forloop
                            
                        NSData *photoData = [NSData dataWithBytes:[cardPublicData getCardHolderPhoto] length:[cardPublicData getCardHolderPhotoLength]];
                        UIImage *pPhotoImage=[[UIImage alloc] initWithData:photoData];
                            
                        NSData *signData = [NSData dataWithBytes:[cardPublicData getHolderSignatureImage] length:[cardPublicData getHolderSignatureImageLength]];
                        UIImage *pSignImage=[[UIImage alloc] initWithData:signData];
                        self.photoImage.image=pPhotoImage;;
                        self.signImage.image=pSignImage;
                        [self.listTableview reloadData];
                        }//if
        });// update UI main queue
}

#pragma mark TableView Delegate and Data source Methods

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 5;
}
- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    switch (section) {
        case 0:
            return @"IDNumber info";
        case 1:
            return @"ModData info";
        case 2:
            return @"NonModData info";
        case 3:
            return @"HomeAddress info";
        case 4:
            return @"WorkAddress info";
    }//switch
    return nil;
}
- (void)tableView:(UITableView *)tableView willDisplayHeaderView:(UIView *)view forSection:(NSInteger)section {
    view.tintColor = [Utils colorFromHexString:@"#BE9647"];
}
- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section; {
    return 32.0;
}
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    switch (section) {
        case 0:
            return pIDNumberKeyarr.count;
        case 1:
            return pModDatakeyarr.count;
        case 2:
            return pNonModDatakeyarr.count;
        case 3:
            return pHomeAddresskeyarr.count;
        case 4:
            return pWorkAddresskeyarr.count;
    }//switch
    return 0 ;
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    static NSString *customTabelId=@"identifierCell";
    CustomTableViewCell *cell=(CustomTableViewCell *)[tableView dequeueReusableCellWithIdentifier:customTabelId];
    if (cell == nil) {
        cell = [[CustomTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:customTabelId];
    }// if
    switch (indexPath.section) {
        case 0:
            cell.keyData.text = [NSString stringWithFormat:@"%@",[pIDNumberKeyarr objectAtIndex:indexPath.row]];
            cell.valueData.text = [NSString stringWithFormat:@"%@",[pIDNumberValuearr objectAtIndex:indexPath.row]];
            break;
        case 1:
            cell.keyData.text = [NSString stringWithFormat:@"%@",[pModDatakeyarr objectAtIndex:indexPath.row]];
            cell.valueData.text = [NSString stringWithFormat:@"%@",[pModDataValuearr objectAtIndex:indexPath.row]];
            break;
        case 2:
            cell.keyData.text =[NSString stringWithFormat:@"%@", [pNonModDatakeyarr objectAtIndex:indexPath.row]];
            cell.valueData.text = [NSString stringWithFormat:@"%@",[pNonModDataValuearr objectAtIndex:indexPath.row]];
            break;
        case 3:
            cell.keyData.text = [NSString stringWithFormat:@"%@",[pHomeAddresskeyarr objectAtIndex:indexPath.row]];
            cell.valueData.text = [NSString stringWithFormat:@"%@",[pHomeAddressValuearr objectAtIndex:indexPath.row]];
            break;
        case 4:
            cell.keyData.text =[NSString stringWithFormat:@"%@", [pWorkAddresskeyarr objectAtIndex:indexPath.row]];
            cell.valueData.text =[NSString stringWithFormat:@"%@", [pWorkAddressValuearr objectAtIndex:indexPath.row]];
            break;
        default:
            break;
    }//switch
    return cell;
}

- (void)goBack {
    [self.navigationController popViewControllerAnimated:YES];
}
#pragma mark didReceiveMemoryWarning
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/ICASwiftApp/ICASwiftApp/NFCPublicDataViewController.swift
//
//  NFCPublicDataViewController.swift
//  ICASwiftApp
//
//  Created by Digital Trust Tech on 06/05/22.
//

import UIKit

class NFCPublicDataViewController: UIViewController,UITableViewDelegate, UITableViewDataSource {
    
    @IBOutlet weak var photoImage: UIImageView!
    @IBOutlet weak var signImage: UIImageView!
    @IBOutlet var listTableview: UITableView!
    @IBOutlet weak var view_Roundedcorners: UIView!
    
    var getpModDataDictionary = NSMutableDictionary()
    var getpNonModDataDictionary = NSMutableDictionary()
    var getpHomeAddressDictionary = NSMutableDictionary()
    var getpWorkAddressDictionary = NSMutableDictionary()
       
    var pIDNumberKeyarr = NSMutableArray()
    var pIDNumberValuearr = NSMutableArray()
    var pModDatakeyarr = NSMutableArray()
    var pModDataValuearr = NSMutableArray()
    var pNonModDatakeyarr = NSMutableArray()
    var pNonModDataValuearr = NSMutableArray()
    var pHomeAddresskeyarr = NSMutableArray()
    var pHomeAddressValuearr = NSMutableArray()
    var pWorkAddresskeyarr = NSMutableArray()
    var pWorkAddressValuearr = NSMutableArray()

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.view_Roundedcorners.cornered_View([.layerMaxXMaxYCorner, .layerMinXMaxYCorner], radius: 20, borderColor: .red, borderWidth: 0)

        photoImage.layer.cornerRadius=8.0
        photoImage.layer.borderWidth=1.0
        photoImage.layer.masksToBounds=true
        photoImage.layer.borderColor=UIColor.clear.cgColor
        photoImage.layer.shadowOffset = CGSize(width: CGFloat(-15), height: CGFloat(20))
        photoImage.layer.shadowRadius=8
        photoImage.layer.shadowOpacity=0.1
               
        signImage.layer.cornerRadius=8.0
        signImage.layer.borderWidth=1.0
        signImage.layer.masksToBounds=true
        signImage.layer.borderColor=UIColor.clear.cgColor
        signImage.layer.shadowOffset = CGSize(width: CGFloat(-15), height: CGFloat(20))
        signImage.layer.shadowRadius=8
        signImage.layer.shadowOpacity=0.1
    }
    

    func getPublicaDataClass(_ cardPublicData:CardPublicData) {
        DispatchQueue.main.asyncAfter(deadline: .now() + 2) {
        //DispatchQueue.main.async {
            self.pIDNumberKeyarr.add("IDNumber")
            self.pIDNumberKeyarr.add("CardNumber")
            self.pIDNumberValuearr.add("\(cardPublicData.getIdNumber())")
            self.pIDNumberValuearr.add("\(cardPublicData.getCardNumber())")
                                
            let parseData = PublicDataParse()
            modifiabledata = cardPublicData.getModifiablePublicData()
            nonModifiabledata = cardPublicData.getNonModifiablePublicData()
            homeaddress = cardPublicData.getHomeAddress()
            workaddress = cardPublicData.getWorkAddress()
                                
            self.getpModDataDictionary = parseData.getModifiablePublicDataDetails(modifiabledata!)
            self.getpNonModDataDictionary = parseData.getNonModifiablePublicDataDetails(nonModifiabledata!)
            self.getpHomeAddressDictionary = parseData.getHomeAddressDetails(homeaddress!)
            self.getpWorkAddressDictionary = parseData.getWorkAddressDetails(workaddress!)
                        
            let isEmpty = self.getpModDataDictionary.count == 0
            if !isEmpty {
                var pModDataKeys: [Any]
                pModDataKeys = self.getpModDataDictionary.allKeys
                for key: String in pModDataKeys as! [String] {
                    let value = self.getpModDataDictionary[key]
                    self.pModDatakeyarr.add(key)
                    self.pModDataValuearr.add(value ?? String())
                }// for loop
                var pNonModDataKeys: [Any]
                pNonModDataKeys = self.getpNonModDataDictionary.allKeys
                for key: String in pNonModDataKeys as! [String] {
                    let value = self.getpNonModDataDictionary[key]
                    self.pNonModDatakeyarr.add(key)
                    self.pNonModDataValuearr.add(value ?? String())
                }// for loop
                var pHomeAddressKeys: [Any]
                pHomeAddressKeys = self.getpHomeAddressDictionary.allKeys
                for key: String in pHomeAddressKeys as! [String] {
                    let value = self.getpHomeAddressDictionary[key]
                    self.pHomeAddresskeyarr.add(key)
                    self.pHomeAddressValuearr.add(value ?? String())
                }// for loop
                var pWorkAddressKeys: [Any]
                pWorkAddressKeys = self.getpWorkAddressDictionary.allKeys
                for key: String in pWorkAddressKeys  as! [String] {
                    let value = self.getpWorkAddressDictionary[key]
                    self.pWorkAddresskeyarr.add(key)
                    self.pWorkAddressValuearr.add(value ?? String())
                }// for loop
                                                        
                let photoData: NSData = NSData(bytes: cardPublicData.getCardHolderPhoto(), length: Int(cardPublicData.getCardHolderPhotoLength()))
                let pPhotoImage = UIImage(data: photoData as Data)
                                    
                let signData: NSData = NSData(bytes: cardPublicData.getHolderSignatureImage(), length: Int(cardPublicData.getHolderSignatureImageLength()))
                let pSignImage = UIImage(data: signData as Data)
                                    
                self.photoImage.image=pPhotoImage
                self.signImage.image=pSignImage
                self.listTableview.reloadData()
            }//if
         } // update UI main queue
    }
    
    // MARK:TableView Delegate and Data source Methods
    func numberOfSections(in tableView: UITableView) -> Int {
        return 5
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        switch section {
        case 0:
            return pIDNumberKeyarr.count
        case 1:
            return pModDatakeyarr.count
        case 2:
            return pNonModDatakeyarr.count
        case 3:
            return pHomeAddresskeyarr.count
        case 4:
            return pWorkAddresskeyarr.count
        default:
            break
        }// switch
        return 0
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        switch section {
        case 0:
            return "IDNumber info"
        case 1:
            return "ModData info"
        case 2:
            return "NonModData info"
        case 3:
            return "HomeAddress info"
        case 4:
            return "WorkAddress info"
        default:
            break
        }// switch
        return nil
    }
    
    func tableView(_ tableView: UITableView, willDisplayHeaderView view: UIView, forSection section: Int) {
        view.tintColor = Utils.color(fromHexString:"#BE9647")
    }
    
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 32.0
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let customTabelId="identifierCell"
        let cell: CustomTableViewCell=(tableView.dequeueReusableCell(withIdentifier: customTabelId) as? CustomTableViewCell)!
        switch indexPath.section {
        case 0:
            cell.keyData.text = pIDNumberKeyarr[indexPath.row] as? String
            cell.valueData.text = pIDNumberValuearr[indexPath.row] as? String
            break
        case 1:
            cell.keyData.text = pModDatakeyarr[indexPath.row] as? String
            cell.valueData.text = pModDataValuearr[indexPath.row] as? String
            break
        case 2:
            cell.keyData.text = pNonModDatakeyarr[indexPath.row] as? String
            cell.valueData.text = pNonModDataValuearr[indexPath.row] as? String
            break
        case 3:
            cell.keyData.text = pHomeAddresskeyarr[indexPath.row] as? String
            cell.valueData.text = pHomeAddressValuearr[indexPath.row] as? String
            break
        case 4:
            cell.keyData.text = pWorkAddresskeyarr[indexPath.row] as? String
            cell.valueData.text = pWorkAddressValuearr[indexPath.row] as? String
            break
        default:
            break
        }// switch
        return cell
    }

    @IBAction func backButtonAction(_ sender: Any) {
        navigationController?.popViewController(animated: true)
    }
}
